### PR TITLE
AyeletK assignment1 solution

### PR DESCRIPTION
**Pull request by ayeletK**
Added new solutions for assignment1 - Q1 and Q2.

### CODE REVIEW

File: ayeletK/assignment1/Q1_solution.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		    char curr_char = str[i];
+		    if (!case_sensitive){
+		        curr_char = tolower(curr_char);
+		    }
+		    curr_str += curr_char;
+		}
+	}
+  // save the last word in the sentence, in case it ends with a word
+	if (!curr_str.empty()){
+	    sort(curr_str.begin(), curr_str.end());
```
Comment (Brusi) [inline ayeletK/assignment1/Q1_solution.cpp]: This 'sort and push' is a code duplication, it appears similarly in lines 55-56.
Such duplications are best avoided; for example, when someone edits this code (let's say, replaces the sort with a different, more efficient function) they can change one occurrence of this and forget to change the other one.

A solution to this is to move the sort to the end of the function, and sort the vector after you are done adding all the words to it.
Another solution is take the sort out of this function completely as I suggested above.

File: ayeletK/assignment1/Q1_solution.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    
+    example:
+    str = "Like a G6"
+    case_sensitive = false
+    ------------------
+    words: <"eikl", "a", "6g">
+*/
+vector<string> split_to_words(string str, bool case_sensitive){
+    vector<string> words;
+    string curr_str = "";
```
Comment (Brusi) [inline ayeletK/assignment1/Q1_solution.cpp]: `string curr_str;` is enough, a string is defaultly initialized to an empty string.

File: ayeletK/assignment1/Q1_solution.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    for (int i = 0; i<str.length() ; ++i){
+    // add new word to words vector when reaching to a punctuation mark
```
Comment (Brusi) [inline ayeletK/assignment1/Q1_solution.cpp]: When a comment is due, I try to make it more 'high-level' than the code.
The reader can see in the code that it "adds an new word to the vector when reaching a punctiation mark".
A more high-level notion can help the reader get better understanding, for example:
```C++
// Reaching a punctuation mark means an end of a word. Add the accumulated word to the vector.
```
See that this comment does not explain only 'what', but also 'why'.

File: ayeletK/assignment1/Q1_solution.cpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    ------------------
+    true
+*/
+bool are_anagrams(string str1, string str2, bool case_sensitive = false){
+    // 	split the string to words
+    vector<string> words1 = split_to_words(str1, case_sensitive);
+	vector<string> words2 = split_to_words(str2, case_sensitive);
+    // 	compare number of words in each
+    if (words1.size() != words2.size()){
+	    return false;
```
Comment (Brusi) [inline ayeletK/assignment1/Q1_solution.cpp]: Funny indentation, please fix :)

File: ayeletK/assignment1/Q1_solution.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	    false otherwise.
+    
+    example:
+    c = "!"
+    ------------------
+    true
+*/
+bool is_punctuation_mark(char c){
+    if (c == ' ' || c == ',' || c == '.' ||
+        c == '!' || c == '?'){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment1/Q1_solution.cpp]: Thanks, I like your way better

File: ayeletK/assignment1/Q1_solution.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	  from the next list: ‘ ‘/’,’/’.’/’!’/’?’
+    @param c - the character to be checked
+    @return true if the character is a punctuation mark,
+	    false otherwise.
+    
+    example:
+    c = "!"
+    ------------------
+    true
+*/
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment1/Q1_solution.cpp]: Note taken!

File: ayeletK/assignment1/Q1_solution.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	vector<string> words2 = split_to_words(str2, case_sensitive);
+    // 	compare number of words in each
+    if (words1.size() != words2.size()){
+	    return false;
+	}
+	// sorting the vectors
+	sort(words1.begin(), words1.end());
+	sort(words2.begin(), words2.end());
+	// compare vectors
+	if (words1 == words2){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment1/Q1_solution.cpp]: note taken

File: ayeletK/assignment1/Q1_solution.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    
+    example:
+    str1 = "Good morning, world"
+    str1 = "doog world mornign!"
+    case_sensitive = false
+    ------------------
+    true
+*/
+bool are_anagrams(string str1, string str2, bool case_sensitive = false){
+    // 	split the string to words
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment1/Q1_solution.cpp]: note taken

File: ayeletK/assignment1/Q1_solution.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+      vector represent a sorted word from the original
+      sentence.
+    
+    example:
+    str = "Like a G6"
+    case_sensitive = false
+    ------------------
+    words: <"eikl", "a", "6g">
+*/
+vector<string> split_to_words(string str, bool case_sensitive){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment1/Q1_solution.cpp]: You make a very good point here. Thanks, I'll make the needed changes!

File: ayeletK/assignment1/Q1_solution.cpp (hunk 1 above)
Comment (ayeletK) [reply ayeletK/assignment1/Q1_solution.cpp]: Note taken. Do you think I can avoid checking the case the sentence ended with a word outside the "for" loop?

File: ayeletK/assignment1/Q2_solution.cpp [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    linked list = 1 -> 2 -> 3 -> 4 -> 5
+    k = 2
+    ------------------
+    4
+*/
+template <typename T>
+Node<T>* get_kth_last_element(Node<T>* head, int k){
+    if (k < 1) return NULL; // k out of bounds
+    // count num of elements in the linked list.
+    Node<T>* curr_node = head;
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment1/Q2_solution.cpp]: You are right. I missed that one

File: ayeletK/assignment1/Q2_solution.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    int size = 0;
+    while (curr_node != NULL){
+        ++size;
+        curr_node = curr_node->next;
+    }
+    if (k > size) return NULL; // k out of bounds
+    // get kth last node
+    curr_node = head;
+    int curr_pos = size;
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment1/Q2_solution.cpp]: It is indeed easy to complicate simple things :)

**Comment by Brusi** [comment]
If you wrote code that tests this, please include it in the commit (maybe we can learn something interesting about it too).

**Comment by ayeletK** [comment]
I added the tests methods.
I know there is a better way to write those tests and I wish to learn.
Any comment will be welcomed :)

**Comment by Brusi** [comment]
More important than how you write your tests, is that you include usage examples and interesting edge cases in the tests, and test all the 'interface' functions.
You checked various case-sensitive and insensitive cases, for both words and sentences, and with special characters. In the list assignment, you also tested the get_size method. Well done!
Additional interesting examples can be: checking 'last_kth' on an empty list (k is zero and non-zero). Also, checking whether two empty words/sentences are anagrams.

Also, it is important that you can actually run these tests to see the code works as you intended.
In this case, I assume a `main()` function that runs all the tests functions and checks everything returns `true` can work (I think this was your intention. If so, you could include this `main` function too :) )

About the best way to write tests: You will have a `testing` session next week; until then, the way you wrote the tests is just fine :)
[One note which you will definitely learn next week: It is best to write tests in a way that you not only know that something failed, but know exactly which test had failed. You will learn how to do it next week].

Good work!
